File: WebsocketServer/WebsocketServer/TypeScriptClient/lib/mgr/MasterSlavesGroupService.ts
namespace WebsocketTSClient
{
    export enum MasterSlavesGroupServiceState
    {
        Idle,
        IsMaster,
        IsSlave
    }

    export class MasterClient
    {
        public clientId: string;
        public masterName: string;
        public isOnline: boolean;        

        constructor(clientId: string, masterName: string, isOnline: boolean)
        {
            this.clientId = clientId;
            this.masterName = masterName;
            this.isOnline = isOnline;
        }

        public static parse(json: any): MasterClient | null
        {
            if (!json || typeof json !== 'object')
            {
                return null;
            }

            const clientId = json.clientId;
            const masterName = json.masterName;
            const isOnline = json.isOnline;

            if (typeof clientId !== 'string' || typeof masterName !== 'string' || typeof isOnline !== 'boolean')
            {
                return null;
            }

            return new MasterClient(clientId, masterName, isOnline);
        }

        public toString(): string
        {
            return `masterName: ${this.masterName}, isOnline: ${this.isOnline}, clientId: ${this.clientId}`;
        }
    }

    export class MasterSlavesGroupService
    {
        /** 服务名称 */
        private static serviceName: string = "MasterSlavesGroupService";
        /** 是否在请求中 */
        private isQuarying: boolean = false;   
        /** 注册为 Listerner */
        public OnRegisteredAsListener: EventHandler<[number]> = new EventHandler<[number]>();
        /** 注销 Listerner */
        public OnUnregisteredFromListener: EventHandler<[number]> = new EventHandler<[number]>();
        /** 注册为 master */
        public OnRegisteredAsMaster: EventHandler<[number]> = new EventHandler<[number]>();
        /** 注销 master */
        public OnUnregisteredFromMaster: EventHandler<[number]> = new EventHandler<[number]>();
        /** 注册为 slave */
        public OnRegisteredAsSlave: EventHandler<[number, MasterClient, any]> = new EventHandler<[number, MasterClient, any]>();
        /** 注销 slave */
        public OnUnregisteredFromSlave: EventHandler<[number]> = new EventHandler<[number]>();
        /** 当服务器 master 集合发生变化 */
        public OnMasterCollectionChanged: EventHandler<[MasterClient]> = new EventHandler<[MasterClient]>();
        /** 获取所有 master */
        public OnGetAllMasters: EventHandler<[number, MasterClient[]]> = new EventHandler<[number, MasterClient[]]>();
        /** 广播消息 */
        public OnBroadcast: EventHandler<[number]> = new EventHandler<[number]>();
        /** 接收到广播消息 */
        public OnRecievedBroadcast: EventHandler<[any]> = new EventHandler<[any]>();

        private state: MasterSlavesGroupServiceState = MasterSlavesGroupServiceState.Idle;

        public GetState(): MasterSlavesGroupServiceState
        {
            return this.state;
        }

        constructor()
        {
            WSBackend.singleton.OnStateChanged.AddListener(this.OnBackendStateChanged);
            WSBackend.singleton.OnNotify.AddListener(this.OnBackendNotify);
        }

        private OnBackendStateChanged = (state: boolean): void =>
        {
            if (!state)
            {
                this.isQuarying = false;
                this.state = MasterSlavesGroupServiceState.Idle;
            }
        }

        private OnBackendNotify = (not: NotifyPack): void =>
        {
            if (not.serviceName === MasterSlavesGroupService.serviceName)
            {
                // 收到服务器 master 集合变化的消息
                if (not.cmd === BackendOps.Notify_OnMasterCollectionChanged)
                {
                    const master: MasterClient | null = MasterClient.parse(not.data);
                    Utility.LogDebug("MasterSlavesGroupService", "Notify Master Collection Changed", master.toString());
                    this.OnMasterCollectionChanged.Trigger(master);
                }
                // 收到广播消息
                if (not.cmd === BackendOps.Cmd_Broadcast)
                {
                    this.OnRecievedBroadcast.Trigger(not.data);
                }
            }
        }

        public RegisterAsListener(): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_RegisterAsListener,
                    null,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.IsMaster : MasterSlavesGroupServiceState.Idle;
                        this.OnRegisteredAsListener.Trigger(errCode);
                    }
                );
            }
        }

        public UnregisterFromListener(): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_UnregisterFromListener,
                    null,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.IsMaster : MasterSlavesGroupServiceState.Idle;
                        this.OnUnregisteredFromListener.Trigger(errCode);
                    }
                );
            }
        }

        public RegisterAsMaster(masterName: string, masterData: any | null): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                var data =
                {
                    masterName: masterName,
                    masterData: masterData
                };
                this.state = MasterSlavesGroupServiceState.IsMaster;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_RegisterAsMaster,
                    data,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.IsMaster : MasterSlavesGroupServiceState.Idle;
                        this.OnRegisteredAsMaster.Trigger(errCode);
                    }
                );
            }
        }

        public UnRegisterFromMaster(): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_UnregisterFromMaster,
                    null,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.Idle : MasterSlavesGroupServiceState.IsMaster;
                        this.OnUnregisteredFromMaster.Trigger(errCode);
                    }
                );
            }
        }

        public RegisterAsSlave(master: MasterClient): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                var data = 
                {
                    masterId: master.clientId
                };
                this.state = MasterSlavesGroupServiceState.IsSlave;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_RegisterAsSlave,
                    data,
                    master,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.IsSlave : MasterSlavesGroupServiceState.Idle;
                        this.OnRegisteredAsSlave.Trigger(errCode, context, data);
                    }
                );
            }
        }

        public UnregisterFromSlave(): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_UnregisterFromSlave,
                    null,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.state = errCode === ErrCode.OK ? MasterSlavesGroupServiceState.Idle : MasterSlavesGroupServiceState.IsSlave;
                        this.OnUnregisteredFromSlave.Trigger(errCode);
                    }
                );
            }
        }

        public GetAllMasters(): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_GetAllMasters,
                    null,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;

                        const result: MasterClient[] = [];
                        const jarr: any[] | null = data["masters"];

                        if (jarr !== null)
                        {
                            jarr.forEach((jobj) =>
                            {
                                var mc = MasterClient.parse(jobj);
                                if (mc !== null)
                                {
                                    result.push(mc);
                                }
                            });
                        }
                        this.OnGetAllMasters.Trigger(errCode, result);
                    }
                );
            }
        }

        public Broadcast(data: any): void
        {
            if (!this.isQuarying && WSBackend.singleton.IsConnected)
            {
                this.isQuarying = true;
                BackendRequest.CreateRetry(
                    MasterSlavesGroupService.serviceName,
                    BackendOps.Cmd_Broadcast,
                    data,
                    null,
                    // resp
                    (errCode: number, data: any, context: any) =>
                    {
                        this.isQuarying = false;
                        this.OnBroadcast.Trigger(errCode);
                    }
                );
            }
        }
    }
}

File: WebSocketUnityClient/Assets/WebSocketClient/Runtime/Scripts/Services/MasterSlavesGroupService.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UIFramework;
using UnityEngine;
using WebSocketClient;

namespace WebSocketClient
{
    public class MasterSlavesGroupService : BackendServiceManagerBase
    {
        public class MasterClient
        {
            public string clientId;          
            public string masterName;
            public bool isOnline;

            public static MasterClient Parse(JToken token)
            {
                if (token == null)
                {
                    return null;
                }
                var cid = JHelper.GetJsonString(token, "clientId");
                var masterName = JHelper.GetJsonString(token, "masterName");
                var isOnline = JHelper.GetJsonBool(token, "isOnline");

                if (string.IsNullOrEmpty(cid))
                {
                    return null;
                }
                return new MasterClient()
                {
                    clientId = cid,
                    masterName = masterName,
                    isOnline = isOnline
                };
            }

            public override string ToString()
            {
                return $"masterName:{masterName}, {isOnline}, clientId:{clientId}";
            }
        }

        public enum ClientState
        {
            Idle,           
            IsMaster,
            IsSlave
        }

        public ClientState clientState { get; private set; } = ClientState.Idle;

        private bool _isQuarying = false;
        private const string serviceName = "MasterSlavesGroupService";

        /// <summary>
        /// 注册成为 master
        /// </summary>
        public event Action<int> OnRegisterAsListener;

        /// <summary>
        /// 注销 master
        /// </summary>
        public event Action<int> OnUnregisteredFromListener;


        /// <summary>
        /// 注册成为 master
        /// </summary>
        public event Action<int> OnRegisteredAsMaster;

        /// <summary>
        /// 注销 master
        /// </summary>
        public event Action<int> OnUnregisteredFromMaster;

        /// <summary>
        /// 注册成为 slave
        /// </summary>
        public event Action<int, JToken> OnRegisteredAsSlave;

        /// <summary>
        /// 注销 slave
        /// </summary>
        public event Action<int> OnUnregisteredFromSlave;

        /// <summary>
        /// 服务器上的 master 集合发生变化
        /// </summary>
        public event Action<MasterClient> OnMasterCollectionChanged;

        /// <summary>
        /// 获取所有的 master
        /// </summary>
        public event Action<int, IEnumerable<MasterClient>> OnGetAllMasters;

        /// <summary>
        /// 广播消息完成
        /// </summary>
        public event Action<int> OnBroadcast;

        /// <summary>
        /// 收到别人发出的广播消息
        /// </summary>
        public event Action<JToken> OnRecievedBroadcast;

        public override async Task Init()
        {
            WSBackend.singleton.OnBackendNotify += Singleton_OnBackendNotify;
            WSBackend.singleton.OnBackendStateChanged += Singleton_OnBackendStateChanged;
            await Task.CompletedTask;
        }

        public override async Task Release()
        {
            WSBackend.singleton.OnBackendNotify -= Singleton_OnBackendNotify;
            WSBackend.singleton.OnBackendStateChanged -= Singleton_OnBackendStateChanged;
            await Task.CompletedTask;
        }

        private void Singleton_OnBackendStateChanged()
        {
            if (WSBackend.singleton.State == WSBackend.WSBackendState.Close)
            {
                // 断线后状态为空闲
                clientState = ClientState.Idle;
                _isQuarying = false;
                Utility.LogDebug("MasterSlavesGroupService", "Connect Closed,become idle state");
            }
        }

        private void Singleton_OnBackendNotify(NotifyPack not)
        {
            if (not.serviceName == MasterSlavesGroupService.serviceName)
            {
                // master collectoion on server changed
                if (not.cmd == BackendOps.Notify_OnMasterCollectionChanged)
                {
                    var masterClient=  MasterClient.Parse(not.data);
                    if (masterClient != null)
                    {
                        Utility.LogDebug("MasterSlavesGroupService", "Notify Master Collection Changed", masterClient);
                        OnMasterCollectionChanged?.Invoke(masterClient);
                    }
                }
                // 收到别人发送的消息
                else if (not.cmd == BackendOps.Cmd_Broadcast)
                {
                    Utility.LogDebug("MasterSlavesGroupService", $"Recieved Broadcast msg {not.data}");
                    OnRecievedBroadcast?.Invoke(not.data);
                }
            }
        }

        /// <summary>
        /// 注册本机作为 Listenr
        /// </summary>
        /// <param name="masterName"></param>
        public async void RegisterAsListener()
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "RegisterAsListener Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_RegisterAsListener, null))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    OnRegisterAsListener?.Invoke(resp.errCode);
                    Utility.LogDebug("MasterSlavesGroupService", "RegisterAsListener End", resp.errCode);                    
                }
            }
        }

        /// <summary>
        /// 注销 Listenr
        /// </summary>
        /// <param name="masterName"></param>
        public async void UnregisterFromListener()
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "UnregisterFromListener Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_UnregisterFromListener, null))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    OnUnregisteredFromListener?.Invoke(resp.errCode);
                    Utility.LogDebug("MasterSlavesGroupService", "UnregisterFromListener End", resp.errCode);
                }
            }
        }

        /// <summary>
        /// 注册本机作为 master
        /// </summary>
        /// <param name="masterName"></param>
        public async void RegisterAsMaster(string masterName, int displayIndex, JToken masterData)
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                clientState = ClientState.IsMaster;
                JObject data = new JObject();
                data.Add("masterName", masterName);
                data.Add("displayIndex", displayIndex);
                data.Add("masterData", masterData);
                Utility.LogDebug("MasterSlavesGroupService", "RegisterAsMaster Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_RegisterAsMaster, data))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    OnRegisteredAsMaster?.Invoke(resp.errCode);
                    Utility.LogDebug("MasterSlavesGroupService", "RegisterAsMaster End", resp.errCode);
                }
            }
        }

        /// <summary>
        /// 注销本机作为 master
        /// </summary>
        public async void UnRegisterFromMaster()
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "UnRegisterFromMaster Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_UnregisterFromMaster, null))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    clientState = resp.errCode == ErrCode.OK ? ClientState.Idle : ClientState.IsMaster;
                    OnUnregisteredFromMaster?.Invoke(resp.errCode);
                    Utility.LogDebug("MasterSlavesGroupService", "UnRegisterFromMaster End", resp.errCode);
                }
            }
        }

        /// <summary>
        /// 注册本机作为 slave
        /// </summary>
        public async void RegisterAsSlave(string masterId)
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                clientState = ClientState.IsSlave;
                JObject data = new JObject();
                data.Add("masterId", masterId);
                Utility.LogDebug("MasterSlavesGroupService", "RegisterAsSlave Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_RegisterAsSlave, data))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    clientState = resp.errCode == ErrCode.OK ? ClientState.IsSlave : ClientState.Idle;
                    OnRegisteredAsSlave?.Invoke(resp.errCode, resp.data);
                    Utility.LogDebug("MasterSlavesGroupService", "RegisterAsSlave End", resp.errCode);
                }
            }
        }

        public async void UnregisterFromSlave()
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "UnregisterFromSlave Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_UnregisterFromSlave, null))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    clientState = resp.errCode == ErrCode.OK ? ClientState.Idle : ClientState.IsSlave;
                    Utility.LogDebug("MasterSlavesGroupService", "UnregisterFromSlave End", resp. errCode);
                    OnUnregisteredFromSlave?.Invoke(resp.errCode);
                }
            }
        }

        public async void GetAllMasters()
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "GetAllMasters Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_GetAllMasters, null))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    List<MasterClient> list = new List<MasterClient>();
                    var arr = JHelper.GetJsonArray(resp.data, "masters");
                    if (arr != null)
                    {
                        foreach (var d in arr)
                        {
                            var master = MasterClient.Parse(d);
                            if (master != null)
                            {
                                list.Add(master);
                            }
                        }
                    }
                    Utility.LogDebug("MasterSlavesGroupService", "GetAllMasters End", $"Get {list.Count} masters");
                    OnGetAllMasters?.Invoke(resp.errCode, list);
                }                
            }
        }

        public async void Broadcast(JToken data)
        {
            if (!_isQuarying && WSBackend.singleton.State == WSBackend.WSBackendState.Open)
            {
                _isQuarying = true;
                Utility.LogDebug("MasterSlavesGroupService", "Broadcast Begin");
                using (var request = new BackendRequestAsync(serviceName, BackendOps.Cmd_GetAllMasters, data))
                {
                    var resp = await request.Request();
                    _isQuarying = false;
                    Utility.LogDebug("MasterSlavesGroupService", "Broadcast End", resp.errCode);
                    OnBroadcast?.Invoke(resp.errCode);
                }
            }
        }
    }
}
